src/hashtable.c

```diff
@@ -0,0 +1,2135 @@
+    if (b->chained && !hashtableIsRehashingPaused(ht)) {
+        /* Rehashing paused also means bucket chain compaction paused. It is
+         * paused while iterating and when a scan callback is running, to be
+         * able to live up to the scan and iterator guarantees. In those cases,
+         * we do the compaction in the scan and iterator code instead. */
+        fillBucketHole(ht, b, pos_in_bucket, table_index);
+    }
+}
+
+/* Initializes the state for an incremental find operation.
+ *
+ * Incremental find can be used to speed up the loading of multiple objects by
+ * utilizing CPU branch predictions to parallelize memory accesses. Initialize
+ * the data for a number of incremental find operations. Then call
+ * hashtableIncrementalFindStep on them in a round-robin order until all of them
+ * are complete. Finally, if necessary, call hashtableIncrementalFindGetResult.
+ */
+void hashtableIncrementalFindInit(hashtableIncrementalFindState *state, hashtable *ht, const void *key) {
+    incrementalFind *data = incrementalFindFromOpaque(state);
+    if (hashtableSize(ht) == 0) {
+        data->state = HASHTABLE_NOT_FOUND;
+    } else {
+        data->state = HASHTABLE_NEXT_BUCKET;
+        data->bucket = NULL;
+        data->hashtable = ht;
+        data->key = key;
+        data->hash = hashKey(ht, key);
+    }
+}
+
+/* Returns 1 if more work is needed, 0 when the done. Call this function
+ * repeatedly until it returns 0. Then use hashtableIncrementalFindGetResult to
+ * fetch the result. */
+int hashtableIncrementalFindStep(hashtableIncrementalFindState *state) {
+    incrementalFind *data = incrementalFindFromOpaque(state);
+    switch (data->state) {
+    case HASHTABLE_CHECK_ENTRY:
+        /* Current entry is prefetched. Now check if it's a match. */
+        {
+            hashtable *ht = data->hashtable;
+            void *entry = data->bucket->entries[data->pos];
+            const void *elem_key = entryGetKey(ht, entry);
+            /* printf("Checking entry at pos %d for key=%p vs elem=%p\n", */
+            /*        data->pos, data->key, elem_key); */
+            if (compareKeys(ht, data->key, elem_key) == 0) {
+                /* It's a match. */
+                data->state = HASHTABLE_FOUND;
+                return 0;
+            }
+            /* No match. Check to next entry in bucket. */
+            data->pos++;
+        }
+        /* fall through */
+    case HASHTABLE_NEXT_ENTRY:
+        /* Current bucket is prefetched. Prefetch next potential
+         * matching entry in the current bucket. */
+        if (data->bucket->presence != 0 && data->pos < numBucketPoitions(data->bucket)) {
+            bucket *b = data->bucket;
+            uint8_t h2 = highBits(data->hash);
+            for (int pos = data->pos; pos < numBucketPoitions(b); pos++) {
+                if (isPositionFilled(b, pos) && b->hashes[pos] == h2) {
+                    /* It's a candidate. */
+                    prefetch(b->entries[pos]);
```

Comment:
Perhaps we can prefetch all entries with the matching hash in parallel, since we have access to all entries simultaneously. In the prev implementation, we had to wait for each entry to be prefetched before accessing its next pointer.

Reply:
It's possible, yes, but it's unlikely that there is more than one matching hash. The probability of mismatch is 1/256 and we have only up to 7 entries. Do you think it will have any impact?

Should we fetch the child bucket too and prefetch matching entries too, even before we check the first bucket's candidate entries? It would make the implementation more complex.

Reply:
I believe your'e right, I'll run some tests to validate it doesn't make a difference.